docs/hello_nf-core/04_make_module.md

```diff
@@ -490,78 +568,147 @@ However, you might want to keep it as a reference for understanding the differen
-    └── tags.yml           # Test tags
+modules/local/cowpy
+├── environment.yml
+├── main.nf
+├── meta.yml
+└── tests
+    └── main.nf.test
 ```
 
-The generated `main.nf` includes all the patterns automatically:
+Each file serves a specific purpose:
 
-```groovy
+- **`main.nf`**: Process definition with all the nf-core patterns built in
+- **`meta.yml`**: Module documentation describing inputs, outputs, and the tool
+- **`environment.yml`**: Conda environment specification for dependencies
+- **`tests/main.nf.test`**: nf-test test cases to validate the module works
+
+!!! tip "Learn more about testing"
+
+    The generated test file uses nf-test, a testing framework for Nextflow pipelines and modules. To learn how to write and run these tests, see the [nf-test side quest](../../side_quests/nf_test/).
+
+The generated `main.nf` includes all the patterns you just learned, plus some additional features:
+
+```groovy title="modules/local/cowpy/main.nf" hl_lines="11 21 22"
 process COWPY {
     tag "$meta.id"
     label 'process_single'
 
     conda "${moduleDir}/environment.yml"
-    container "..."
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/YOUR-TOOL-HERE':
+        'biocontainers/YOUR-TOOL-HERE' }"
 
     input:
-    tuple val(meta), path(input_file)      // Metadata tuples ✓
+    tuple val(meta), path(input)        // Pattern 1: Metadata tuples ✓
 
     output:
-    tuple val(meta), path("${prefix}.*"), emit: output  // Metadata propagation ✓
-    path "versions.yml"                   , emit: versions
+    tuple val(meta), path("*"), emit: output
+    path "versions.yml"           , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
 
     script:
-    def args = task.ext.args ?: ''                    // ext.args pattern ✓
-    def prefix = task.ext.prefix ?: "${meta.id}"    // ext.prefix pattern ✓
+    def args = task.ext.args ?: ''              // Pattern 2: ext.args ✓
+    def prefix = task.ext.prefix ?: "${meta.id}"  // Pattern 3: ext.prefix ✓
+
     """
-    # TODO: Add your command here
-    cowpy $args < $input_file > ${prefix}.txt
+    // Add your tool command here
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
-        cowpy: \$(cowpy --version | sed 's/cowpy //')
+        cowpy: \$(cowpy --version)
+    END_VERSIONS
+    """
+
+    stub:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+
+    """
+    echo $args
+    touch ${prefix}.txt
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        cowpy: \$(cowpy --version)
     END_VERSIONS
     """
 }
 ```
 
-You fill in the command logic and the module is ready to test!
+Notice how all three patterns you applied manually are already there!
+The template also includes several additional nf-core conventions that we didn't cover in the hands-on section:
```

Comment:
This part is confusing because it's not clear if we have to do something about these extra bits. Is the idea that the module is runnable as is so don't worry about it? If so, is it submittable to nf-core as-is, or would we have to do additional work to fill these bits out before attempting a submission? 

The next section covers the container stuff but not any of the rest.

Reply:
OK, extended the tweaking a bit to address this